dlib/matrix/matrix_fft.h

```diff
@@ -9,837 +9,142 @@
-    inline matrix<std::complex<double>,0,1> ifft(const matrix<std::complex<double>,0,1>& data)
-    {
-        return call_mkl_fft(data, true) / data.size();
-    }
-    inline matrix<std::complex<double>,1,0> fft (const matrix<std::complex<double>,1,0>& data)
-    {
-        return call_mkl_fft(data, false);
-    }
-    inline matrix<std::complex<double>,1,0> ifft(const matrix<std::complex<double>,1,0>& data)
-    {
-        return call_mkl_fft(data, true) / data.size();
-    }
-    inline matrix<std::complex<double> > fft (const matrix<std::complex<double> >& data)
-    {
-        return call_mkl_fft(data, false);
-    }
-    inline matrix<std::complex<double> > ifft(const matrix<std::complex<double> >& data)
+    template < typename T, long NR, long NC, typename MM, typename L >
+    void ifft_inplace (matrix<std::complex<T>,NR,NC,MM,L>& data)
     {
-        return call_mkl_fft(data, true) / data.size();
+        kiss_fft({(int)data.nr(), (int)data.nc()}, &data(0,0), &data(0,0), true);
     }
 
-    inline void fft_inplace (matrix<std::complex<double>,0,1>& data)
-    {
-        call_mkl_fft_inplace(data, false);
-    }
-    inline void ifft_inplace(matrix<std::complex<double>,0,1>& data)
-    {
-        call_mkl_fft_inplace(data, true);
-    }
-    inline void fft_inplace (matrix<std::complex<double>,1,0>& data)
-    {
-        call_mkl_fft_inplace(data, false);
-    }
-    inline void ifft_inplace(matrix<std::complex<double>,1,0>& data)
+// ----------------------------------------------------------------------------------------
+    
+    template <typename EXP>
+    matrix<std::complex<typename EXP::type>> fftr (const matrix_exp<EXP>& data)
     {
-        call_mkl_fft_inplace(data, true);
+        // You have to give a complex matrix
+        COMPILE_TIME_ASSERT(std::is_floating_point<typename EXP::type>::value);
+        DLIB_ASSERT(data.nc() % 2 == 0, "last dimension needs to be even otherwise ifftr(fftr(data)) won't have matching dimensions : " << data.nc());
+        matrix<typename EXP::type> in(data);
+        matrix<std::complex<typename EXP::type>> out(in.nr(), in.nc()/2+1);
+        kiss_fftr({(int)in.nr(), (int)in.nc()}, &in(0,0), &out(0,0));
+        return out;
     }
 
-    inline void fft_inplace (matrix<std::complex<double> >& data)
-    {
-        call_mkl_fft_inplace(data, false);
-    }
-    inline void ifft_inplace(matrix<std::complex<double> >& data)
+// ----------------------------------------------------------------------------------------
+    
+    template <typename EXP>
+    matrix<typename remove_complex<typename EXP::type>::type> ifftr (const matrix_exp<EXP>& data)
     {
-        call_mkl_fft_inplace(data, true);
+        // You have to give a complex matrix
+        COMPILE_TIME_ASSERT(is_complex<typename EXP::type>::value);        
+        matrix<typename remove_complex<typename EXP::type>::type> out;
+        if (data.size() == 0)
+            return out;
+        matrix<typename EXP::type> in = data;
```

Comment:
This copy is not needd when the input type is already this type of matrix.  So there should be some overload or something to avoid this when it's not needed.

Reply:
yeah, i was going to add some overloads for ```dlib::matrix``` and ```std::vector```. For ```std::vector```, i really don't want to do unecessary copies of the form ```std::vector - > dlib::matrix_exp -> dlib::matrix```

Reply:
Can you think of any other useful overloads?

Reply:
Also, what do you think the return type of the overload for ```std::vector``` should be? ```std::vector``` or ```dlib::matrix``` ?